content/blog/posts/o3de-interactive-tutorials-prototyping-2022.md

```diff
@@ -0,0 +1,40 @@
+
+Thus, I risked creating duplicate entities or potentially breaking the tutorial by trying to edit entities which didn't yet exist when the 'simulate' option was selection. I knew I could solve the problem by condensing the tutorial into three giant steps, but that seemed inefficient and also unhelpful for the user. I decided to create a 'simulate last step' option which displayed the completed result of the tutorial as a reference for the user.
+In the meantime, I helped move the pyside_utils.py file from EditorPythonTestTools to the QtForPython gem. The InteractiveTutorials gem uses pyside_util.py, so the move would make InteractiveTutorials usable from any project, not just AutomatedTesting. After attending a SIG-Testing meeting we updated the automated testing scripts to import pyside_utils from the new location. The experience impressed upon me the importance of communication and clarity, especially when a codebase as large as O3DE's.
+
+Finally, as part of my stretch goal, I worked on improving the highlights UI using Qt. My focus was to improve the visibility of small widgets. I came up with 3 new highlighting options, depending on the size of the widget and the color of the background. First, the content of the widget itself could be shaded either blue or red. Also, the parent widget could also be highlighted. Finally, I explored adding a blinking / pulsing effect.
+
+### Looking Back, Looking Forward
+One Amazon Principle that has stuck with me is Learn and Be Curious. I saw an example of this principle in the inception of the InteractiveTutorials gem during one developer’s PD week and in my mentor’s suggestions for how the gem could be improved, and I attempted to emulate it by trying to solve problems that I'm still not sure are solvable.
+
+For future work on automation, I would suggest working bottom-up rather than top-down. I would attempt to automate a very simple tutorial, such as Create an Entity, and to develop a general purpose solution, rather than jumping out of the gate and trying to automate the entire Rigid Bodies tutorial on a first pass. I would also consider using an abstract data type to memoize or keep track of the current-step-index and the last-simulated-step index, since many problems which rely on consecutive if-else gates can also be solved dynamically. Finally, instead of keeping track of the last-simulated-step-index, I might consider using some of the Python Tests in O3DE to determine which entities have already appeared and been edited. If a test returns False, then the entities it tracks must not have been created yet. An advantage of O3DE being open source is that I can continue to learn and be curious about this project long after the summer’s over. This may sound trite, but I've genuinely enjoyed my summer at Amazon, and I'm so thankful not only for the opportunity to write here, but also to have been part of a team that encouraged and challenged me to work with rigor and attention to detail.
```

Comment:
```suggestion
For future work on automation, I would suggest working bottom-up rather than top-down. I would attempt to automate a very simple tutorial, such as Create an Entity, and to develop a general purpose solution, rather than jumping out of the gate and trying to automate the entire Rigid Bodies tutorial on a first pass. I would also consider using an abstract data type to memorize or keep track of the current-step-index and the last-simulated-step index, since many problems which rely on consecutive if-else gates can also be solved dynamically. Finally, instead of keeping track of the last-simulated-step-index, I might consider using some of the Python Tests in O3DE to determine which entities have already appeared and been edited. If a test returns False, then the entities it tracks must not have been created yet. An advantage of O3DE being open source is that I can continue to learn and be curious about this project long after the summer’s over. This may sound trite, but I've genuinely enjoyed my summer at Amazon, and I'm so thankful not only for the opportunity to write here, but also to have been part of a team that encouraged and challenged me to work with rigor and attention to detail.
```